build-system/replace-secrets.js

```diff
@@ -0,0 +1,96 @@
+/**
+ * Copyright 2020 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Replaces redacted secret environment variables in an app
+ * directory's `redacted.env` and outputs `.env` with the decrypted values of
+ * the secrets.
+ */
+
+const fs = require('fs');
+const path = require('path');
+const YAML = require('yaml');
+
+const OUTPUT_ENV_FILE = '.env';
+const REDACTED_ENV_FILE = 'redacted.env';
+const CLOUD_BUILD_FILE = 'cloud_build.yaml';
+
+/**
+ * Reads the Cloud Build config file to identify secrets in the environment.
+ *
+ * @param {string} appDir path to app directory.
+ * @return {Array<string>} list of environment variable names.
+ */
+function identifySecrets(appDir) {
+  const yamlFile = path.join(appDir, CLOUD_BUILD_FILE);
+  try {
+    const yamlContents = fs.readFileSync(yamlFile).toString('utf8');
+    const config = YAML.parse(yamlContents);
+    const secrets = Object.keys(config.secrets[0].secretEnv);
+    console.log(`Identified the following secret env keys: ${secrets}`);
+    return secrets;
+  } catch (e) {
+    // Handle the case where parsing secrets from the cloud build file fails.
+    console.warn('No secret keys were found in cloud_build.yaml.');
+  }
+
+  return [];
+}
+
+/**
+ * Reads the redacted env file and creates a `.env` file including secrets.
+ *
+ * @param {string} appDir path to app directory.
+ */
+function replaceSecrets(appDir) {
+  const envFile = path.join(appDir, REDACTED_ENV_FILE);
+  const envFileContents = fs.readFileSync(envFile).toString('utf8');
+
+  const replacedContents = envFileContents
+    .split('\n')
+    .map(line => {
+      for (const secret of identifySecrets(appDir)) {
+        if (line.startsWith(`${secret}=`)) {
+          const secretVal = process.env[secret];
+          console.log(
+            `Replacing value of ${secret} with "${secretVal.substr(0, 3)}..."`
+          );
+          return `${secret}=${secretVal}`;
+        }
+      }
+
+      return line;
+    })
+    .join('\n');
+
+  fs.writeFileSync(path.join(appDir, OUTPUT_ENV_FILE), replacedContents);
+}
+
+if (require.main === module) {
```

Comment:
This'll make @jridgewell sad :(

Reply:
What's the canonical version of this?

Reply:
Resolving since the `NODE_ENV` check rendered this outdated.

Reply:
This does make me sad. Do this in entry points, not imported code.

Reply:
Can you clarify what you mean by entry points?

Also, this code should realistically never be imported (I only added the exports bit out of habit tbh). Should I drop the exports?

Reply:
Entry points are the script that you actually invoke: `node foo.js` has `foo.js` as an entry point. If `foo.js` imports `bar.js`, `bar.js` should never touch the process invocation.

If there's no exports, is this the file that's being invoked?

Reply:
Yes, this is run by Cloud Build (and only in Cloud Build, never locally) and is never imported by another script.

Reply:
Then the if statement should be unnecessary. These sometimes get moved into `bin` folders, to signify that they're scripts to execute and not modules to import.

Reply:
Done and done